src/XrdCeph/XrdCephBulkAioRead.cc

```diff
@@ -0,0 +1,160 @@
+    librados::AioCompletion *cmpl = std::get<1>(tup.second);
+    cmpl->release();
+    delete op;
+  }
+  operations.clear();
+};
+
+int bulkAioRead::addRequest(std::string objname, char* out_buf, size_t size, off64_t offset) {
+  ceph::bufferlist *bl;
+  int *retval;
+  librados::ObjectReadOperation *rop;
+
+  try {
+    bl = new ceph::bufferlist();
+  } catch (std::bad_alloc&) {
+    log_func((char*)"Can not allocate buffer for read (%lu, %lu)", offset, size);
+    return -ENOMEM;
+  }
+
+  try {
+    retval = new int;
+  } catch (std::bad_alloc&) {
+    log_func((char*)"Can not allocate int for read's retval (%lu, %lu)", offset, size);
+    delete bl;
+    return -ENOMEM;
+  }
+
+  auto tup = operations.find(objname);
+  if (operations.end() == tup) {
+    try {
+      rop = new librados::ObjectReadOperation;
+    } catch (std::bad_alloc&) {
+      log_func((char*)"Can not allocate int for read object op (%lu, %lu)", offset, size);
+      delete bl;
+      delete retval;
+      return -ENOMEM;
+    }
+
+    librados::AioCompletion *cmpl = librados::Rados::aio_create_completion();
+    if (0 == cmpl) {
+      log_func((char*)"Can not create completion for read (%lu, %lu)", offset, size);
+      delete bl;
+      delete retval;
+      delete rop;
+      return -ENOMEM;
+    }
+    operations.insert( {objname, std::make_pair(rop, cmpl)} );
+  } else {
+    rop = std::get<0>(tup->second);
+  }
+  buffers.push_back( std::make_tuple(bl, out_buf, retval) );
+
+  rop->read(offset, size, bl, retval);
+  return 0;
+};
+
+void bulkAioRead::wait_for_complete() {
+  std::string obj_name;
+  librados::AioCompletion* cmpl;
+  librados::ObjectReadOperation* op;
+  for (auto const& tup: operations) {
+    obj_name = tup.first;
+    op = std::get<0>(tup.second);
+    cmpl = std::get<1>(tup.second);
+    context->aio_operate(obj_name, cmpl, op, 0);
+  }
+
+  for (auto const& tup: operations) {
+    cmpl = std::get<1>(tup.second);
+    cmpl->wait_for_complete();
+  }
+};
+
+ssize_t bulkAioRead::get_results() {
+  ceph::bufferlist* bl;
+  char *buf;
+  int *rc;
+  ssize_t res = 0;
+  for (ReadOpData i: buffers) {
+    bl = std::get<0>(i);
+    buf = std::get<1>(i);
+    rc = std::get<2>(i);
+    if (*rc < 0) {
+      log_func((char*)"One of the reads failed with rc %d", rc);
+      return *rc;
+    }
+    bl->begin().copy(bl->length(), buf);
+    res += bl->length();
+  }
+  return res;
+};
+
+int bulkAioRead::read(void* out_buf, size_t req_size, off64_t offset) {
+
+  size_t end_block, buf_pos, chunk_len, chunk_start, req_len;
+  //We are using start_block in printf with %s, so we have to use int.
+  //Hopefully, our file's sizes are not too big...
+  unsigned int start_block;
+  char *buf_ptr;
+  //16 bytes for object hex number, 1 for dot and 1 for null-terminator
+  char object_suffix[18];
+  std::string obj_name;
+  int rc;
+
+  req_len = req_size;
+  start_block = offset / block_size;
+  end_block = (offset + req_len - 1) / block_size;
```

Comment:
What if req_len == 0?

Reply:
Since `offset ` and `req_len` are `size_t`, i.e. unsigned, the result in braces will be equal to max `size_t` value. And `end_block` will be huge:
```
(gdb) p start_block
$13 = 1
(gdb) p end_block
$14 = 4398046511103
(gdb)
```

So we must not proceed if req_len is <= 0.

Reply:
An alternative might be:
```
while (req_len > 0) {
  size_t const chunk_len = std::min(req_len, block_size);
  assert(chunk_len > 0);

  // skipped obj_name

  int rc = addRequest(obj_name, buf_ptr, chunk_len, chunk_start);

  // skipped increments
  req_len -= chunk_len;
}
```
Then `end_block` is redundant, and you're also assured that you're not passing `chunk_len==0` to `addRequest`.

Reply:
It maybe just me, but I feel a little bit uncomfortable subtracting from unsigned and then comparing to zero.
Though yes, it is another option, and we can switch to it.

Reply:
You could do `assert(chunk_len <= req_len)` (or equivalent) just before the subtraction.

Reply:
So should we switch to the `req_len > 0` condition?

I think that this `req_len > 0` condition makes exit condition less obvious, but it may be just me.

Reply:
Since `req_len == req_size - buf_pos`, you could compute it as such where necessary, and then use `buf_pos < req_size` as the (more obvious?) condition, dropping req_len altogether.  You could similarly drop `buf_ptr`, first validating the range [`buf_pos`, `buf_pos + chunk_len`), then computing it from the validated `buf_pos`:
```
char *const cbuf = (char *) out_buf; // just for readability later
while (buf_pos < req_size) {
  size_t const chunk_len = std::min(req_size - buf_pos,
                                    object_size - chunk_start);

  // We'll never add a zero-length request.
  assert(chunk_len > 0);

  // The end of the request won't exceed the caller's buffer.
  // This will also be the new value of buf_pos,
  // so the next buf_pos is already validated.
  size_t const buf_end = buf_pos + chunk_len;
  assert(buf_end <= req_size);

  int rc;
  rc = addRequest(start_block, cbuf + buf_pos, chunk_len, chunk_start);
  if (rc < 0) { ... }

  // The next request abuts this one in the destination buffer.
  buf_pos = buf_end;

  // The next read is from the start of the next block.
  chunk_start = 0;
  start_block++;
}
```

Reply:
I think, it is the same "less obvious" exit condition. In the current version loop variable `start_block` is increase by a fixed value on every iteration, that's why it is clear from the beginning how many iterations will be performed.

In case of size-related conditions it is more complicated: a value that is subtracted from the `req_len`/added to `buf_pos` can change on every iteration. 

If we are to choose between `req_len > 0` and `buf_pos < req_size`, I'd prefer the first one. That one was implemented.